### PR TITLE
TST: check internal Categorical

### LINKED ISSUE #13076: TST: assert_series/frame not comparing for categoricals
xref https://github.com/dask/partd/pull/7

```
In [1]: df1 = DataFrame({'A' : Series(list('aba')).astype('category', categories=list('ab'), ordered=True)})

In [2]: df2 = DataFrame({'A' : Series(list('aba')).astype('category', categories=list('ba'), ordered=True)})

In [3]: tm.assert_frame_equal(df1, df2)

In [11]: df2 = DataFrame({'A' : Series(list('aba')).astype('category', categories=list('ab'), ordered=False)})

In [12]: tm.assert_frame_equal(df1, df2)

In [13]: df2 = DataFrame({'A' : Series(list('aba')).astype('category', categories=list('ba'), ordered=False)})

In [14]: tm.assert_frame_equal(df1, df2)
```

all of these should fail as the ordered categoricals don't compare equal in the categories, and ordered is not being compared (nor categories)

Furthermore the ``assert_series_equal`` fail for the same.

``assert_categorical_equal`` DOES the right thing, but not being called here. (not that this *should* say ``categories`` are not equal?)
```
In [17]: tm.assert_categorical_equal(df1.A.values, df2.A.values)
AssertionError: Index are different

Index values are different (100.0 %)
[left]:  Index([u'a', u'b'], dtype='object')
[right]: Index([u'b', u'a'], dtype='object')
```

### PR DESCRIPTION
**Pull request by sinhrks**
- [x] closes #13076
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`

### CODE REVIEW

File: pandas/util/testing.py [hunk 1]
```diff
@@ -685,8 +687,8 @@ def _check_types(l, r, obj='Index'):
             assert_class_equal(left, right, exact=exact, obj=obj)
             assert_attr_equal('dtype', l, r, obj=obj)
             # allow string-like to have different inferred_types
-            if l.inferred_type in ('string', 'unicode'):
-                assertIn(r.inferred_type, ('string', 'unicode'))
+            if l.inferred_type in ('string', 'unicode', 'mixed'):
```
Comment (jreback) [inline pandas/util/testing.py]: hmm, you sure about this? 'mixed' could be lots of things, even an integer & string will trigger.

Comment (sinhrks) [inline pandas/util/testing.py]: Ah, this is no longer required. Removed (originally tried to handle ``.categories`` contains ``NaN``).

File: pandas/tests/frame/test_reshape.py [hunk 2]
```diff
@@ -253,9 +253,11 @@ def test_unstack_fill(self):
         # By default missing values will be NaN
         result = data.unstack()
         expected = DataFrame({'a': pd.Categorical(list('axa'),
-                                                  categories=list('abc')),
```
Comment (jreback) [inline pandas/tests/frame/test_reshape.py]: so we explicity force these to ordered? (or is it starting with an ordered cat)? this seems odd.

File: pandas/tests/test_categorical.py [hunk 3]
```diff
@@ -556,28 +556,35 @@ def test_categories_none(self):
     def test_describe(self):
         # string type
         desc = self.factor.describe()
+        self.assertTrue(self.factor.ordered)
+        exp_index = pd.CategoricalIndex(['a', 'b', 'c'], ordered=True,
+                                        name='categories')
         expected = DataFrame({'counts': [3, 2, 3],
                               'freqs': [3 / 8., 2 / 8., 3 / 8.]},
-                             index=pd.CategoricalIndex(['a', 'b', 'c'],
-                                                       name='categories'))
```
Comment (jreback) [inline pandas/tests/test_categorical.py]: are we explicity forcing this to ordered? (not against it, more curious where/why we did this)

File: pandas/tests/test_categorical.py [hunk 4]
```diff
@@ -4083,13 +4098,14 @@ def f():
...
             c.set_categories(["a", "b", np.nan], rename=True, inplace=True)
+
         c[0] = np.nan
         df = pd.DataFrame({"cats": c, "vals": [1, 2, 3]})
         df_exp = pd.DataFrame({"cats": Categorical(["a", "b", "a"]),
                                "vals": [1, 2, 3]})
 
         res = df.fillna("a")
-        tm.assert_frame_equal(res, df_exp)
+        tm.assert_frame_equal(res, df_exp, check_categorical=False)
```
Comment (jreback) [inline pandas/tests/test_categorical.py]:
```
In [11]: df.fillna('a').cats
Out[11]: 
0    a
1    b
2    a
Name: cats, dtype: category
Categories (2, object): [a, b]
```

does this not match?

**Comment by jreback** [comment]
thanks @sinhrks

**Pull request closed without merging**